lib/simulator/ScopeStates.js

```diff
@@ -0,0 +1,145 @@
+import {
+  ScopePhases
+} from './ScopePhases';
+
+
+function illegalTransition(state, target) {
+  throw new Error(`illegal transition: ${state.name} -> ${target}`);
+}
+
+/**
+ * A representation of a scopes state.
+ */
+export class ScopeState {
+
+  /**
+   * @param {string} name
+   * @param {number} phase
+   * @param {object} [transitions]
+   * @param {ScopeState} [transitions.start]
+   * @param {ScopeState} [transitions.compensate]
+   * @param {ScopeState} [transitions.complete]
+   * @param {ScopeState} [transitions.destroy]
+   * @param {ScopeState} [transitions.fail]
+   * @param {ScopeState} [transitions.terminate]
+   */
+  constructor(name, phase, {
+    start,
+    compensate,
+    complete,
+    destroy,
+    fail,
+    terminate
+  } = {}) {
+    this.name = name;
+    this.phase = phase;
+
+    this._start = start;
+    this._compensate = compensate;
+    this._complete = complete;
+    this._destroy = destroy;
+    this._fail = fail;
+    this._terminate = terminate;
+  }
+
+  /**
+   * @param {number} phase
+   * @return {boolean}
+   */
+  matchesPhase(phase) {
+    return (this.phase & phase) !== 0;
+  }
```

Comment:
So this is the part where my eyes are wide open and I need to find [this](https://github.com/bpmn-io/bpmn-js-token-simulation/pull/101/files#diff-bd5bf0c64a375e2e6502dd9ac58aaf0c9d765dcec5740f4d3303662898201aafR11) to understand why it's written like that. How about if we structure this like below:

```javascript
NOT_DEAD = [  ACTIVATED, RUNNING, ENDING ];

matchesPhase(phaseOrPhases) {
  const phases = isArray(phaseOrPhases) ? phaseOrPhases : [ phaseOrPhases ];
  return phases.some(phase => this.phase === phase);
}
```

?

Reply:
Perhaps there could be a comment with explanation why it works like that. I see that [later in the code](https://github.com/bpmn-io/bpmn-js-token-simulation/pull/101/files#diff-288e93a31248fe0b480308968969acb8bb525759fd8c49181c35bda0712831d3R96) we combine the different states.

Reply:
What you're writing would actually work just fine. 

This check is potentially executed a lot of times; hence I wanted to go for something that is efficient (which usually conflicts with _understandable_). Bit-wise is often times a solution to exactly what [you suggest](https://github.com/bpmn-io/bpmn-js-token-simulation/pull/101#discussion_r786656228) albeit a lot more compact.

If you use bit-wise operators you consider a number in binary form:

````
var a = 0x00000000001;
var b = 0x11111111111;
```

You can `&` or `|` them to combine them, by bit wise `and` or `or`:

```
    a = 0x00000000001;
    b = 0x00000000010;
var c = 0x00000000000; // a & b
var d = 0x00000000011; // a | b
```

Bit-wise operations allow me to easily compose state by making each bit represent a particular piece of information. I can test if `x` matches one of these with simple bit-wise arithmetics (rather than complex collection logic):

```
(a | b) & x !== 0 // if x is either a, b, or the combination of both
```

Reply:
I do dislike the overall term `ScopePhases` and renamed it to `ScopeTraits` on the branch.

Reply:
I like the solution but agree that it's harder to understand for us mortal developers as we rarely use this kind of functionality. Maybe a comment about _why_ (you mentioned performance) it was implemented that way would be helpful.

Reply:
What kind of comment would help you?

I do regard this as being internal API; as a user you should not need to know about this anyway (cf. human-understandable wrappers in `Scope`).

Reply:
Just a comment about the reason for using bitwise. That reason being performance. Just a suggestion. I know this code is unlikely to be encountered by any of us.